File: security-platform/agents/java/src/main/java/com/securityplatform/agent/Redactor.java
package com.securityplatform.agent;

import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * Redacts PII from data structures.
 */
public class Redactor {
    private static final String DEFAULT_REPLACEMENT = "***REDACTED***";
    private List<RedactionRule> rules;

    public Redactor(List<RedactionRule> rules) {
        this.rules = rules != null ? rules : List.of();
    }

    public Object redact(Object data) {
        if (data == null) {
            return null;
        }

        if (data instanceof Map) {
            return redactMap((Map<String, Object>) data);
        } else if (data instanceof List) {
            return redactList((List<?>) data);
        } else if (data instanceof String) {
            return redactString((String) data);
        }

        return data;
    }

    private Map<String, Object> redactMap(Map<String, Object> map) {
        Map<String, Object> result = new java.util.HashMap<>(map);
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            String key = entry.getKey();
            Object value = entry.getValue();

            // Check if key matches any redaction rule
            boolean shouldRedact = false;
            for (RedactionRule rule : rules) {
                if (rule.getField() != null && rule.getField().equals(key)) {
                    shouldRedact = true;
                    break;
                }
                if (rule.getPattern() != null && Pattern.compile(rule.getPattern(), Pattern.CASE_INSENSITIVE).matcher(key).find()) {
                    shouldRedact = true;
                    break;
                }
            }

            if (shouldRedact) {
                result.put(key, DEFAULT_REPLACEMENT);
            } else {
                result.put(key, redact(value));
            }
        }
        return result;
    }

    private List<?> redactList(List<?> list) {
        return list.stream().map(this::redact).collect(java.util.stream.Collectors.toList());
    }

    private String redactString(String str) {
        for (RedactionRule rule : rules) {
            if (rule.getPattern() != null) {
                Pattern pattern = Pattern.compile(rule.getPattern(), Pattern.CASE_INSENSITIVE);
                String replacement = rule.getReplacement() != null ? rule.getReplacement() : DEFAULT_REPLACEMENT;
                str = pattern.matcher(str).replaceAll(replacement);
            }
        }
        return str;
    }
}

File: security-platform/agents/go/redactor.go
package agent

import (
	"reflect"
	"regexp"
	"strings"
)

const defaultReplacement = "***REDACTED***"

// Redactor redacts PII from data structures.
type Redactor struct {
	rules []RedactionRule
}

// NewRedactor creates a new redactor with the given rules.
func NewRedactor(rules []RedactionRule) *Redactor {
	return &Redactor{rules: rules}
}

// Redact redacts PII from the given data.
func (r *Redactor) Redact(data interface{}) interface{} {
	if data == nil {
		return nil
	}

	val := reflect.ValueOf(data)
	switch val.Kind() {
	case reflect.Map:
		return r.redactMap(data)
	case reflect.Slice, reflect.Array:
		return r.redactSlice(data)
	case reflect.String:
		return r.redactString(data.(string))
	default:
		return data
	}
}

func (r *Redactor) redactMap(data interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	val := reflect.ValueOf(data)

	for _, key := range val.MapKeys() {
		keyStr := key.String()
		value := val.MapIndex(key).Interface()

		shouldRedact := false
		for _, rule := range r.rules {
			if rule.Field != "" && strings.EqualFold(rule.Field, keyStr) {
				shouldRedact = true
				break
			}
			if rule.Pattern != "" {
				matched, err := regexp.MatchString("(?i)"+rule.Pattern, keyStr)
				if err == nil && matched {
					shouldRedact = true
					break
				}
			}
		}

		if shouldRedact {
			replacement := defaultReplacement
			for _, rule := range r.rules {
				if rule.Field == keyStr || (rule.Pattern != "" && r.matchesPattern(rule.Pattern, keyStr)) {
					if rule.Replacement != "" {
						replacement = rule.Replacement
					}
					break
				}
			}
			result[keyStr] = replacement
		} else {
			result[keyStr] = r.Redact(value)
		}
	}

	return result
}

func (r *Redactor) redactSlice(data interface{}) []interface{} {
	val := reflect.ValueOf(data)
	result := make([]interface{}, val.Len())

	for i := 0; i < val.Len(); i++ {
		result[i] = r.Redact(val.Index(i).Interface())
	}

	return result
}

func (r *Redactor) redactString(str string) string {
	result := str
	for _, rule := range r.rules {
		if rule.Pattern != "" {
			replacement := rule.Replacement
			if replacement == "" {
				replacement = defaultReplacement
			}
			re := regexp.MustCompile("(?i)" + rule.Pattern)
			result = re.ReplaceAllString(result, replacement)
		}
	}
	return result
}

func (r *Redactor) matchesPattern(pattern, str string) bool {
	matched, err := regexp.MatchString("(?i)"+pattern, str)
	return err == nil && matched
}
